### ISSUE TITLE
[Analysis] Needed improvements in runs table for `get-pipelines` and other queries

### ISSUE THREAD

**Issue by theSaarco** [mlrun/mlrun#4524]
## TL;DR
The following actions are needed:
1. Create an index on the `runs` table that contains both `(id, project)` fields.
2. Create another index on the `runs_labels` (and any other labels table) on the `(name, value)` fields, to speed up label-based queries.
3. Need to start reducing the size of the `runs` table, and the associated `runs_labels`. Currently the long-run setup still has information on a run that started in `2023-07-24 11:50:11.664`, which is more than 3 months ago. We can (and should) limit it, since it tracks execution jobs which are mostly interesting when they happen and a little-bit afterwards. Knowledge about a run that happened 3 months ago is not very interesting, and the user will most likely not be able to retrieve it anyway unless knowing exactly the workflow-id/run-id to look for. 

## Background
The long-run setup currently (5/11) has 363,683 records in the `runs` table. Most of them part of the same project (`load-project-0`):
```
+----------------+----------+
| project        | count(*) |
+----------------+----------+
| load-project-0 |   351195 |
| fs-project     |    12487 |
| log-project    |        1 |
+----------------+----------+
```
There are 2,922,879 records in the `runs_labels` table, which are divided to various labels that we stick on runs. The important one for our purposes is the `workflow` label which is put on every run that is a step in a workflow:
```
+-----------------------------+----------+
| name                        | count(*) |
+-----------------------------+----------+
| feature-set                 |     2493 |
| host                        |   360356 |
| job-type                    |     4866 |
| kind                        |   362733 |
| mlrun-auth-key              |   127255 |
| mlrun/client_python_version |   361917 |
| mlrun/client_version        |   361917 |
| mlrun/job                   |    16122 |
| mlrun/runner-pod            |   234661 |
| mlrun/schedule-name         |   128071 |
| owner                       |   362733 |
| v3io_user                   |   362721 |
| workflow                    |   237034 |
+-----------------------------+----------+
```
When doing a `get-pipeline` API call, it translates into this query on the DB (in `format_summary_from_kfp_run`):
```python
# enrich DAG with mlrun run info
runs = run_db.list_runs(project=project, labels=f"workflow={run_id}")
```
We're getting runs from a given project that have a specific workflow label. Nothing wrong with it conceptually.

## DB analysis
The query above translates into the following SQL:
```sql
SELECT runs.id AS runs_id, runs.uid AS runs_uid, runs.project AS runs_project, runs.name AS runs_name, runs.iteration AS runs_iteration, runs.state AS runs_state, runs.body AS runs_body, runs.start_time AS runs_start_time, runs.requested_logs AS runs_requested_logs, runs.updated AS runs_updated 
FROM runs JOIN (
    SELECT runs_labels.id AS id, runs_labels.name AS name, runs_labels.value AS value, runs_labels.parent AS parent 
    FROM runs_labels 
    WHERE runs_labels.name = 'workflow' AND runs_labels.value = '0ba46347-f520-4568-9014-a1c5a2f2a493') AS labels 
    ON runs.id = labels.parent 
WHERE runs.project = 'load-project-0'
```
Which when executed directly on the DB takes a little less than 3 seconds:
```
100 rows in set (2.78 sec)
```
3 seconds is a relatively long time. I thought some of the issue is that the query does subquery rather than a simple join, so I tried to run the following query instead:
```sql
SELECT runs.id AS runs_id, runs.uid AS runs_uid, runs.project AS runs_project, runs.name AS runs_name, runs.iteration AS runs_iteration, runs.state AS runs_state, runs.body AS runs_body, runs.start_time AS runs_start_time, runs.requested_logs AS runs_requested_logs, runs.updated AS runs_updated
FROM runs 
JOIN runs_labels ON runs.id = runs_labels.parent
WHERE  runs_labels.name = 'workflow' AND runs_labels.value = '0ba46347-f520-4568-9014-a1c5a2f2a493' AND runs.project = 'load-project-0'
```
It didn't improve anything at all -
```
100 rows in set (2.87 sec)
```

Next, I thought maybe the DB has some issue with the join complexity, and dropping the join and using a subquery instead with an `in` clause may help?
```sql
SELECT runs.id AS runs_id, runs.uid AS runs_uid, runs.project AS runs_project, runs.name AS runs_name, runs.iteration AS runs_iteration, runs.state AS runs_state, runs.body AS runs_body, runs.start_time AS runs_start_time, runs.requested_logs AS runs_requested_logs, runs.updated AS runs_updated
FROM runs
WHERE runs.project = 'load-project-0' and runs.id in (
    select parent from runs_labels where runs_labels.name = 'workflow' AND runs_labels.value = '0ba46347-f520-4568-9014-a1c5a2f2a493'
)
```
Guess what?
```
100 rows in set (2.92 sec)
```
So the problem is not in how we join the tables. 
Then I thought - the `project` predicate has no impact at all (since all the runs from a workflow are from the same project), so what if I remove it?
```sql
SELECT runs.id AS runs_id, runs.uid AS runs_uid, runs.project AS runs_project, runs.name AS runs_name, runs.iteration AS runs_iteration, runs.state AS runs_state, runs.body AS runs_body, runs.start_time AS runs_start_time, runs.requested_logs AS runs_requested_logs, runs.updated AS runs_updated 
FROM runs JOIN (
    SELECT runs_labels.id AS id, runs_labels.name AS name, runs_labels.value AS value, runs_labels.parent AS parent 
    FROM runs_labels 
    WHERE runs_labels.name = 'workflow' AND runs_labels.value = '0ba46347-f520-4568-9014-a1c5a2f2a493') AS labels 
    ON runs.id = labels.parent 
```

```
100 rows in set (0.87 sec)
```
Aha! Still about 1 second which is not amazing, but 3 times better than before. So, it was the time for the DB to explain itself:

```sql
mysql> explain <query with project predicate>
+----+-------------+-------------+------------+--------+----------------------------------------+-----------------+---------+---------------------+--------+----------+-------------+
| id | select_type | table       | partitions | type   | possible_keys                          | key             | key_len | ref                 | rows   | filtered | Extra       |
+----+-------------+-------------+------------+--------+----------------------------------------+-----------------+---------+---------------------+--------+----------+-------------+
|  1 | SIMPLE      | runs        | NULL       | ALL    | PRIMARY                                | NULL            | NULL    | NULL                | 290087 |    10.00 | Using where |
|  1 | SIMPLE      | runs_labels | NULL       | eq_ref | _runs_labels_uc,_runs_labels_parent_fk | _runs_labels_uc | 773     | const,mlrun.runs.id |      1 |    10.00 | Using where |
+----+-------------+-------------+------------+--------+----------------------------------------+-----------------+---------+---------------------+--------+----------+-------------+
2 rows in set, 1 warning (0.00 sec)

mysql> explain <query without project predicate>
+----+-------------+-------------+------------+--------+----------------------------------------+-----------------+---------+--------------------------+--------+----------+------------------------------------+
| id | select_type | table       | partitions | type   | possible_keys                          | key             | key_len | ref                      | rows   | filtered | Extra                              |
+----+-------------+-------------+------------+--------+----------------------------------------+-----------------+---------+--------------------------+--------+----------+------------------------------------+
|  1 | SIMPLE      | runs_labels | NULL       | ref    | _runs_labels_uc,_runs_labels_parent_fk | _runs_labels_uc | 768     | const                    | 396404 |    10.00 | Using index condition; Using where |
|  1 | SIMPLE      | runs        | NULL       | eq_ref | PRIMARY                                | PRIMARY         | 4       | mlrun.runs_labels.parent |      1 |   100.00 | NULL                               |
+----+-------------+-------------+------------+--------+----------------------------------------+-----------------+---------+--------------------------+--------+----------+------------------------------------+
2 rows in set, 1 warning (0.00 sec)
```

What happens when filtering on `project` - the DB doesn't have any index with that field in it that it can use (there is a unique constraint which is essentially an index, but it's on `(uid, project, iteration)` and doesn't include `id` in it), so it **scans the full table** and filters it based on the project - which doesn't do much since most records are of the same project, then going on to filter the results based on the join with the labels table. Never mind that the DB could have been smarter in that case, it works with what it has...

When querying without the `project`, the DB is using the `PRIMARY` index on the `runs` table (which is on the `id` that is the join field with the labels table), and it can also push the query on the labels to be first, and filter it using an index. However, even in that case it is just `Using index condition` meaning that it uses the index for initial filtering but then needs to still do some query on the rows. That's because the index on the `labels` table is created on `(name, parent)` so it can filter based on name (reducing the result to 237034 rows, rather than ~3M), but then it still needs to filter based on value. Then it takes the resulting 100 rows and retrieves them based on `id` from the `runs` table.

I'm pretty sure (couldn't test it since I don't want to touch the long-run setup DB and create indexes on it) that if we create an index on the `lables` table that indexes `(name, value)` - this query will become much, much faster, as it will filter the results from the labels to 100 rows very efficiently. Also, adding an index on the `runs` table that filters on `(id, project)` will allow the DB to filter on labels first, then filter the `runs` table. It will also be useful for other queries since almost all of our queries on runs involve the `project` as well.

The end result are the points written in the beginning of this note.
